### ISSUE TITLE
datafusion-cli should trap interrupt signal and abort execution

### ISSUE THREAD

**Issue by Jimexist** [apache/arrow-datafusion#1279]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

datafusion-cli should trap interrupt signal and abort execution

**Describe the solution you'd like**

using rustyline signal handling and kill async operation instead of exiting

**Describe alternatives you've considered**

**Additional context**

**Comment by Jefffrey** [comment]
I tried to take an initial shot at this: https://github.com/Jefffrey/arrow-datafusion/commit/4761d169af2f08764af81f56aa1dab128a960167

But have run into a roadblock, where in the `exec_and_print()` function it can potentially block whilst printing, especially for a very large result set.

```rust
async fn exec_and_print(
    ctx: &mut SessionContext,
    print_options: &PrintOptions,
    sql: String,
) -> Result<()> {
    let now = Instant::now();
    let df = ctx.sql(&sql).await?;
    let results = df.collect().await?;
    print_options.print_batches(&results, now)?; // can block here!

    Ok(())
}
```

Since it is blocking, you can't abort the async function whilst it executes. Even spawning it in a separate task with `spawn_blocking()` doesn't look like it'll help, as though it would return control to let you enter more commands when you interrupt with CTRL-C, eventually that synchronous blocking print function will complete anyway and print onto the console (even though you interrupted it).

Perhaps could spawn it in a separate thread and kill it, rather than trying to relying on async cancellation?

**Comment by Jefffrey** [comment]
That could be possible @qrilka, though I believe might require changes in arrow-rs as I think that's where the iterative print code is

Definitely something worth exploring

**Comment by qrilka** [comment]
@Jefffrey my first thought was about having "cancellation points" after/before every batch i.e. in `print_batches` itself. But I guess that would be too coarse-grained and it looks like ideally we need `AsyncWrite` implemented for every printer...

**Comment by Jefffrey** [comment]
Yeah that might be too coarse-grained and I think would vary depending on size of batches? So could still have cases where user hits `CTRL-C` and still has a batch print afterwards, messing with their prompt.

I'll see if I have time to look into this a bit more, though I haven't made any progress on this since my initial stab at it.

**Comment by Jefffrey** [comment]
Closed by #8739
